### PR TITLE
Implement Channel Upgrade Cancel

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #4403

This PR incorporates the recent spec changes into ChannelUpgradeCancel

- Existing tests were uncommented and updated so they are passing with the changes.
- `isAuthorizedUpgrader` will always return true for now, and will be implemented in a follow up.
- `WriteUpgradeCancelChannel` method signature updated to accept an `ErrorReceipt`. 

Note: `restoreChannel` in the spec includes the writing of an error receipt, however our implementation just reverts channel state and deletes upgrade info. The additional call to `SetUpgradeErrorReceipt` ensures the receipt is written along with the restore in `ChannelUpgradeCancel`

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 1]
```diff
@@ -1110,188 +1110,240 @@ func (suite *KeeperTestSuite) TestWriteUpgradeOpenChannel() {
...
+		malleate func()
+		expError error
+	}{
+		{
+			name:     "success with flush complete",
+			malleate: func() {},
+			expError: nil,
+		},
+		{
+			name: "success without flush complete",
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade_test.go:1133]:
```suggestion
			name: "success with flushing state",
```

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 2]
```diff
@@ -1110,188 +1110,240 @@ func (suite *KeeperTestSuite) TestWriteUpgradeOpenChannel() {
...
-// 				var ok bool
-// 				errorReceipt, ok = suite.chainB.GetSimApp().IBCKeeper.ChannelKeeper.GetUpgradeErrorReceipt(suite.chainB.GetContext(), path.EndpointB.ChannelConfig.PortID, path.EndpointB.ChannelID)
-// 				suite.Require().True(ok)
+	tests := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			name:     "success with flush complete",
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade_test.go:1128]:
```suggestion
			name:     "success with flush complete state",
```

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 3]
```diff
@@ -1110,188 +1110,240 @@ func (suite *KeeperTestSuite) TestWriteUpgradeOpenChannel() {
...
+			name: "connection not found",
+			malleate: func() {
+				channel := path.EndpointA.GetChannel()
+				channel.ConnectionHops = []string{"connection-100"}
+				path.EndpointA.SetChannel(channel)
+			},
+			expError: connectiontypes.ErrConnectionNotFound,
+		},
+		{
+			name: "connection not found",
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade_test.go:1185]: duplicate case I think?

File: modules/core/04-channel/keeper/upgrade_test.go [hunk 4]
```diff
@@ -1110,188 +1110,240 @@ func (suite *KeeperTestSuite) TestWriteUpgradeOpenChannel() {
...
+			if expPass {
+				suite.Require().NoError(err)
+			} else {
+				suite.Require().ErrorIs(err, tc.expError)
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestWriteUpgradeCancelChannel() {
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade_test.go:1275]: maybe an issue to refactor this to table tests if possible? what do you think?

**Review by damiannolan — APPROVED** [review]
Happy to move forward with this and address anything else in follow ups! 
Reviewed together synchronously and cross ref'd with spec. 

Just left some notes in this review. 

LGTM!

File: modules/core/04-channel/keeper/upgrade.go [hunk 5]
```diff
@@ -580,35 +591,35 @@ func (k Keeper) ChanUpgradeCancel(ctx sdk.Context, portID, channelID string, err
...
-	counterpartySequence := errorReceipt.Sequence
-	if counterpartySequence < currentSequence {
-		return errorsmod.Wrapf(types.ErrInvalidUpgradeSequence, "error receipt sequence (%d) must be greater than or equal to current sequence (%d)", counterpartySequence, currentSequence)
-	}
-
 	return nil
 }
 
+// isAuthorizedUpgrader checks if the sender is authorized to cancel the upgrade.
+func (Keeper) isAuthorizedUpgrader(signer string) bool {
```
Comment (DimitrisJim) [inline modules/core/04-channel/keeper/upgrade.go:598]: do you remember if we have an issue for this function? A quick glance at issue list didn't yield anything

Comment (damiannolan) [reply modules/core/04-channel/keeper/upgrade.go:598]: aye good call, we need to open one still https://github.com/cosmos/ibc-go/pull/4404#discussion_r1300125982

Comment (damiannolan) [reply modules/core/04-channel/keeper/upgrade.go:598]: https://github.com/cosmos/ibc-go/issues/4420 done

**Pull request merged**